faiss/impl/NSG.h

```diff
@@ -0,0 +1,198 @@
+#include <faiss/impl/AuxIndexStructures.h>
+#include <faiss/impl/FaissAssert.h>
+#include <faiss/utils/Heap.h>
+#include <faiss/utils/random.h>
+
+namespace faiss {
+
+/** Implementation of the Navigating Spreading-out Graph (NSG)
+ * datastructure.
+ *
+ * Fast Approximate Nearest Neighbor Search With The
+ * Navigating Spreading-out Graph
+ *
+ *  Cong Fu, Chao Xiang, Changxu Wang, Deng Cai, VLDB 2019
+ *
+ * This implmentation is heavily influenced by the NSG
+ * implementation by ZJULearning Group
+ * (https://github.com/zjulearning/nsg)
+ *
+ * The NSG object stores only the neighbor link structure, see
+ * IndexNSG.h for the full index object.
+ */
+
+struct DistanceComputer; // from AuxIndexStructures
+struct Neighbor;
+struct Node;
+
+namespace nsg {
+
+/***********************************************************
+ * Graph structure to store a graph.
+ *
+ * It is represented by an adjacency matrix `data`, where
+ * data[i, j] is the j-th neighbor of node i.
+ ***********************************************************/
+
+template <class node_t>
+struct Graph {
+    node_t* data;    // the flattened adjacency matrix
+    int K;           // nb of neighbors per node
+    int N;           // total nb of nodes
+    bool own_fields; // the underlying data owned by itself or not
+
+    // construct from a known graph
+    Graph(node_t* data, int N, int K)
+            : data(data), N(N), K(K), own_fields(false) {}
+
+    // construct an empty graph
+    // NOTE: the newly allocated data needs to be destroyed at destruction time
+    Graph(int N, int K) : N(N), K(K), own_fields(true) {
+        data = new node_t[N * K];
+    }
+
+    // release the allocated memory if needed
+    ~Graph() {
+        if (own_fields) {
+            delete[] data;
+        }
+    }
+
+    // access the j-th neighbor of node i
+    inline node_t at(int i, int j) const {
+        return data[i * K + j];
+    }
+
+    // access the j-th neighbor of node i by reference
+    inline node_t& at(int i, int j) {
+        return data[i * K + j];
+    }
+};
+
+DistanceComputer* storage_distance_computer(const Index* storage);
+
+} // namespace nsg
+
+struct NSG {
+    /// internal storage of vectors (32 bits: this is expensive)
+    using storage_idx_t = int;
+
+    /// Faiss results are 64-bit
+    using idx_t = Index::idx_t;
+
+    // It needs to be smaller than 0
+    static const int EMPTY_ID = -1;
+
+    // nb of nodes
+    int ntotal;
+
+    // construction-time parameters
+    int R; // nb of neighbors per node
+    int L; // length of the search path at construction time
+    int C; // candidate pool size at construction time
+
+    // search-time parameters
+    int search_L; // length of the search path
+
+    // enterpoint
+    int enterpoint;
+
+    // the built graph structure
+    std::shared_ptr<nsg::Graph<int>> final_graph;
+
+    // NSG is built or not
+    bool is_built;
+
+    // random generator
+    mutable RandomGenerator rng;
```

Comment:
Is it feasible to make this a local variable? I would like to avoid having mutable fields.

Reply:
Get rid of the mutable variable now.